File: src/algorithms/sorting/exchange/circle-sort/sources/CircleSort.java
public class CircleSort {
    public static int[] circleSort(int[] inputArray) { // @step:initialize
        int[] sortedArray = inputArray.clone(); // @step:initialize
        int arrayLength = sortedArray.length; // @step:initialize

        // Repeat full passes until no swaps occur
        boolean swapped = true;
        while (swapped) {
            swapped = circleSortPass(sortedArray, 0, arrayLength - 1);
        }

        return sortedArray; // @step:complete
    }

    private static boolean circleSortPass(int[] sortedArray, int leftIndex, int rightIndex) {
        if (leftIndex >= rightIndex) {
            return false;
        }

        boolean swapped = false;
        int low = leftIndex;
        int high = rightIndex;

        while (low < high) {
            // @step:compare
            if (sortedArray[low] > sortedArray[high]) { // @step:compare
                // @step:swap
                int temporaryValue = sortedArray[low]; // @step:swap
                sortedArray[low] = sortedArray[high]; // @step:swap
                sortedArray[high] = temporaryValue; // @step:swap
                swapped = true;
            }
            low++;
            high--;
        }

        // If the midpoint element is reached (odd-length segment), compare with one above
        if (low == high) {
            if (sortedArray[low] > sortedArray[high + 1]) { // @step:compare
                // @step:swap
                int temporaryValue = sortedArray[low]; // @step:swap
                sortedArray[low] = sortedArray[high + 1]; // @step:swap
                sortedArray[high + 1] = temporaryValue; // @step:swap
                swapped = true;
            }
        }

        int midpoint = (leftIndex + rightIndex) / 2;
        boolean leftSwapped = circleSortPass(sortedArray, leftIndex, midpoint);
        boolean rightSwapped = circleSortPass(sortedArray, midpoint + 1, rightIndex);

        return swapped || leftSwapped || rightSwapped;
    }
}

File: src/algorithms/sorting/exchange/circle-sort/sources/circle-sort.py
def circle_sort(input_array: list[int]) -> list[int]:  # @step:initialize
    sorted_array = input_array.copy()  # @step:initialize
    array_length = len(sorted_array)  # @step:initialize

    # Repeat full passes until no swaps occur
    swapped = True
    while swapped:
        swapped = _circle_sort_pass(sorted_array, 0, array_length - 1)

    return sorted_array  # @step:complete


def _circle_sort_pass(sorted_array: list[int], left_index: int, right_index: int) -> bool:
    if left_index >= right_index:
        return False

    swapped = False
    low = left_index
    high = right_index

    while low < high:
        # @step:compare
        if sorted_array[low] > sorted_array[high]:  # @step:compare
            # @step:swap
            temporary_value = sorted_array[low]  # @step:swap
            sorted_array[low] = sorted_array[high]  # @step:swap
            sorted_array[high] = temporary_value  # @step:swap
            swapped = True
        low += 1
        high -= 1

    # If the midpoint element is reached (odd-length segment), compare it with one above
    if low == high:
        if sorted_array[low] > sorted_array[high + 1]:  # @step:compare
            # @step:swap
            temporary_value = sorted_array[low]  # @step:swap
            sorted_array[low] = sorted_array[high + 1]  # @step:swap
            sorted_array[high + 1] = temporary_value  # @step:swap
            swapped = True

    midpoint = (left_index + right_index) // 2
    left_swapped = _circle_sort_pass(sorted_array, left_index, midpoint)
    right_swapped = _circle_sort_pass(sorted_array, midpoint + 1, right_index)

    return swapped or left_swapped or right_swapped
